pkg/batchscheduler/volcano/volcano_scheduler.go

```diff
@@ -0,0 +1,294 @@
+}
+
+func GetPluginName() string {
+	return "volcano"
+}
+
+func (v *VolcanoBatchScheduler) Name() string {
+	return GetPluginName()
+}
+
+func (v *VolcanoBatchScheduler) ShouldSchedule(app *v1beta1.SparkApplication) bool {
+
+	checkScheduler := func(scheduler *string) bool {
+		if scheduler != nil && *scheduler == v.Name() {
+			return true
+		}
+		return false
+	}
+
+	if app.Spec.Mode == v1beta1.ClientMode {
+		return checkScheduler(app.Spec.Executor.SchedulerName)
+	}
+	if app.Spec.Mode == v1beta1.ClusterMode {
+		return checkScheduler(app.Spec.Executor.SchedulerName) && checkScheduler(app.Spec.Driver.SchedulerName)
+	}
+
+	glog.Warningf("Unsupported Spark application mode %s, abandon schedule via volcano.", app.Spec.Mode)
+	return false
+}
+
+func (v *VolcanoBatchScheduler) PatchApplicationPod(pod *corev1.Pod, app *v1beta1.SparkApplication) []util.PatchOperation {
+	var operations []util.PatchOperation
+	if v.ShouldSchedule(app) {
+		// Patch only executor pods
+		if app.Spec.Mode == v1beta1.ClientMode {
+			if util.IsExecutorPod(pod) {
+				patchOperation := v.addVolcanoAnnotation(pod, app)
+				if patchOperation != nil {
+					operations = append(operations, *patchOperation)
+				}
+			}
+			//Patch both driver and executor pods
+		} else if app.Spec.Mode == v1beta1.ClusterMode {
+			if util.IsDriverPod(pod) || util.IsExecutorPod(pod) {
+				patchOperation := v.addVolcanoAnnotation(pod, app)
+				if patchOperation != nil {
+					operations = append(operations, *patchOperation)
+				}
+			}
+		}
+	}
+	return operations
+}
+
+func (v *VolcanoBatchScheduler) addVolcanoAnnotation(pod *corev1.Pod, app *v1beta1.SparkApplication) *util.PatchOperation {
+	path := "/metadata/annotations"
+	var value interface{}
+	if len(pod.Annotations) != 0 {
+		if _, ok := pod.Annotations[v1alpha2.GroupNameAnnotationKey]; ok {
+			return nil
+		}
+		pod.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(app)
+		value = pod.Annotations
+	} else {
+		value = map[string]string{v1alpha2.GroupNameAnnotationKey: v.getAppPodGroupName(app)}
+	}
+	return &util.PatchOperation{Op: "replace", Path: path, Value: value}
+}
+
+func (v *VolcanoBatchScheduler) BeforeApplicationSubmission(app *v1beta1.SparkApplication) error {
+	if app.Spec.Mode == v1beta1.ClientMode {
+		return v.syncPodGroupInClientMode(app)
+	} else if app.Spec.Mode == v1beta1.ClusterMode {
+		return v.syncPodGroupInClusterMode(app)
+	}
+	return nil
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupInClientMode(app *v1beta1.SparkApplication) error {
+	//Only executor resource will be considered.
+	return v.syncPodGroup(app, 1, getExecutorRequestResource(app))
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupInClusterMode(app *v1beta1.SparkApplication) error {
+	totalResource := sumResourceList([]corev1.ResourceList{getExecutorRequestResource(app), getDriverRequestResource(app)})
+	//NOTE: In cluster mode, the initial size of PodGroup is set to 1 in order to schedule driver pod first.
```

Comment:
Are the driver pod and executors pods considered to be in the same group? If so, should the minSize of the group be 2 (1 driver pod plus one executor pod)?

Reply:
The Minresource stands for the Min pods that volcano will start the podgroup. if it sets to 2, the driver will never been started, since there is not coming any executor pods.